File: rust/crates/forge-db/src/event_repository.rs
//! Event repository — persistence for append-only `events` audit log.

use std::collections::HashMap;

use rusqlite::{params, params_from_iter, types::Value, OptionalExtension};
use uuid::Uuid;

use crate::{Db, DbError};

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Event {
    pub id: String,
    pub timestamp: String,
    pub event_type: String,
    pub entity_type: String,
    pub entity_id: String,
    pub payload: String,
    pub metadata: Option<HashMap<String, String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct EventQuery {
    pub event_type: Option<String>,
    pub entity_type: Option<String>,
    pub entity_id: Option<String>,
    pub since: Option<String>,
    pub until: Option<String>,
    pub cursor: String,
    pub limit: i64,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct EventPage {
    pub events: Vec<Event>,
    pub next_cursor: String,
}

fn now_rfc3339() -> String {
    crate::now_rfc3339()
}

fn nullable_string(value: &str) -> Option<&str> {
    if value.is_empty() {
        None
    } else {
        Some(value)
    }
}

fn scan_event_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Event> {
    let payload_json: Option<String> = row.get(5)?;
    let metadata_json: Option<String> = row.get(6)?;
    let metadata = match metadata_json {
        Some(value) => serde_json::from_str::<HashMap<String, String>>(&value).ok(),
        None => None,
    };

    Ok(Event {
        id: row.get(0)?,
        timestamp: row.get(1)?,
        event_type: row.get(2)?,
        entity_type: row.get(3)?,
        entity_id: row.get(4)?,
        payload: payload_json.unwrap_or_default(),
        metadata,
    })
}

pub struct EventRepository<'a> {
    db: &'a Db,
}

impl<'a> EventRepository<'a> {
    pub fn new(db: &'a Db) -> Self {
        Self { db }
    }

    /// Append validates required fields then inserts the event.
    pub fn append(&self, event: &mut Event) -> Result<(), DbError> {
        if event.event_type.trim().is_empty()
            || event.entity_type.trim().is_empty()
            || event.entity_id.trim().is_empty()
        {
            return Err(DbError::InvalidEvent);
        }
        self.create(event)
    }

    /// Create inserts an event and applies Go-compatible defaults.
    pub fn create(&self, event: &mut Event) -> Result<(), DbError> {
        if event.event_type.trim().is_empty() {
            return Err(DbError::Validation("event type is required".into()));
        }
        if event.entity_type.trim().is_empty() {
            return Err(DbError::Validation("event entity type is required".into()));
        }
        if event.entity_id.trim().is_empty() {
            return Err(DbError::Validation("event entity id is required".into()));
        }

        if event.id.is_empty() {
            event.id = Uuid::new_v4().to_string();
        }
        if event.timestamp.is_empty() {
            event.timestamp = now_rfc3339();
        }

        let metadata_json: Option<String> = match &event.metadata {
            Some(value) => Some(serde_json::to_string(value).map_err(|err| {
                DbError::Validation(format!("failed to marshal metadata: {err}"))
            })?),
            None => None,
        };

        self.db.conn().execute(
            "INSERT INTO events (
                id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
            ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
            params![
                event.id,
                event.timestamp,
                event.event_type,
                event.entity_type,
                event.entity_id,
                nullable_string(&event.payload),
                metadata_json,
            ],
        )?;

        Ok(())
    }

    pub fn get(&self, id: &str) -> Result<Event, DbError> {
        let row = self
            .db
            .conn()
            .query_row(
                "SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
                 FROM events
                 WHERE id = ?1",
                params![id],
                scan_event_row,
            )
            .optional()?;
        row.ok_or(DbError::EventNotFound)
    }

    /// Query with cursor-based pagination. Sort order is timestamp asc, id asc.
    pub fn query(&self, q: EventQuery) -> Result<EventPage, DbError> {
        let limit = if q.limit <= 0 { 100 } else { q.limit };
        let mut query = String::from(
            "SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
             FROM events
             WHERE 1=1",
        );
        let mut args: Vec<Value> = Vec::new();

        if let Some(event_type) = q.event_type {
            query.push_str(" AND type = ?");
            args.push(Value::from(event_type));
        }
        if let Some(entity_type) = q.entity_type {
            query.push_str(" AND entity_type = ?");
            args.push(Value::from(entity_type));
        }
        if let Some(entity_id) = q.entity_id {
            query.push_str(" AND entity_id = ?");
            args.push(Value::from(entity_id));
        }
        if let Some(since) = q.since {
            query.push_str(" AND timestamp >= ?");
            args.push(Value::from(since));
        }
        if let Some(until) = q.until {
            query.push_str(" AND timestamp < ?");
            args.push(Value::from(until));
        }
        if !q.cursor.is_empty() {
            query
                .push_str(" AND (timestamp, id) > (SELECT timestamp, id FROM events WHERE id = ?)");
            args.push(Value::from(q.cursor));
        }

        query.push_str(" ORDER BY timestamp, id LIMIT ?");
        args.push(Value::from(limit + 1));

        let mut stmt = self.db.conn().prepare(&query)?;
        let rows = stmt.query_map(params_from_iter(args.iter()), scan_event_row)?;
        let mut events = Vec::new();
        for row in rows {
            events.push(row?);
        }

        if (events.len() as i64) > limit {
            let next_cursor = events[(limit - 1) as usize].id.clone();
            events.truncate(limit as usize);
            return Ok(EventPage {
                events,
                next_cursor,
            });
        }

        Ok(EventPage {
            events,
            next_cursor: String::new(),
        })
    }

    pub fn list_by_entity(
        &self,
        entity_type: &str,
        entity_id: &str,
        limit: i64,
    ) -> Result<Vec<Event>, DbError> {
        let limit = if limit <= 0 { 100 } else { limit };
        let mut stmt = self.db.conn().prepare(
            "SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
             FROM events
             WHERE entity_type = ?1 AND entity_id = ?2
             ORDER BY timestamp
             LIMIT ?3",
        )?;
        let rows = stmt.query_map(params![entity_type, entity_id, limit], scan_event_row)?;
        let mut events = Vec::new();
        for row in rows {
            events.push(row?);
        }
        Ok(events)
    }

    pub fn count(&self) -> Result<i64, DbError> {
        let count: i64 = self
            .db
            .conn()
            .query_row("SELECT COUNT(*) FROM events", [], |row| row.get(0))?;
        Ok(count)
    }

    pub fn oldest_timestamp(&self) -> Result<Option<String>, DbError> {
        let value: Option<String> = self
            .db
            .conn()
            .query_row("SELECT MIN(timestamp) FROM events", [], |row| row.get(0))
            .optional()?
            .flatten();
        Ok(value)
    }

    pub fn delete_older_than(&self, before: &str, limit: i64) -> Result<i64, DbError> {
        let limit = if limit <= 0 { 1000 } else { limit };
        let rows = self.db.conn().execute(
            "DELETE FROM events WHERE id IN (
                SELECT id FROM events WHERE timestamp < ?1 ORDER BY timestamp LIMIT ?2
            )",
            params![before, limit],
        )?;
        Ok(rows as i64)
    }

    pub fn delete_excess(&self, max_count: i64, limit: i64) -> Result<i64, DbError> {
        if max_count <= 0 {
            return Ok(0);
        }
        let limit = if limit <= 0 { 1000 } else { limit };
        let total = self.count()?;
        let excess = total - max_count;
        if excess <= 0 {
            return Ok(0);
        }
        let delete_count = excess.min(limit);
        let rows = self.db.conn().execute(
            "DELETE FROM events WHERE id IN (
                SELECT id FROM events ORDER BY timestamp LIMIT ?1
            )",
            params![delete_count],
        )?;
        Ok(rows as i64)
    }

    pub fn list_older_than(&self, before: &str, limit: i64) -> Result<Vec<Event>, DbError> {
        let limit = if limit <= 0 { 1000 } else { limit };
        let mut stmt = self.db.conn().prepare(
            "SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
             FROM events
             WHERE timestamp < ?1
             ORDER BY timestamp
             LIMIT ?2",
        )?;
        let rows = stmt.query_map(params![before, limit], scan_event_row)?;
        let mut events = Vec::new();
        for row in rows {
            events.push(row?);
        }
        Ok(events)
    }

    pub fn list_oldest(&self, limit: i64) -> Result<Vec<Event>, DbError> {
        let limit = if limit <= 0 { 1000 } else { limit };
        let mut stmt = self.db.conn().prepare(
            "SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
             FROM events
             ORDER BY timestamp
             LIMIT ?1",
        )?;
        let rows = stmt.query_map(params![limit], scan_event_row)?;
        let mut events = Vec::new();
        for row in rows {
            events.push(row?);
        }
        Ok(events)
    }

    pub fn delete_by_ids(&self, ids: &[String]) -> Result<i64, DbError> {
        if ids.is_empty() {
            return Ok(0);
        }

        let placeholders = std::iter::repeat("?")
            .take(ids.len())
            .collect::<Vec<_>>()
            .join(",");
        let query = format!("DELETE FROM events WHERE id IN ({placeholders})");

        let values = ids
            .iter()
            .map(|id| Value::from(id.clone()))
            .collect::<Vec<_>>();

        let rows = self
            .db
            .conn()
            .execute(&query, params_from_iter(values.iter()))?;
        Ok(rows as i64)
    }
}

File: old/go/internal/db/event_repository.go
// Package db provides SQLite database access for Forge.
package db

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/tOgg1/forge/internal/models"
)

// Event repository errors.
var (
	ErrEventNotFound = errors.New("event not found")
	ErrInvalidEvent  = errors.New("invalid event")
)

// EventRepository handles event persistence.
type EventRepository struct {
	db *DB
}

type eventExecer interface {
	ExecContext(context.Context, string, ...any) (sql.Result, error)
}

// NewEventRepository creates a new EventRepository.
func NewEventRepository(db *DB) *EventRepository {
	return &EventRepository{db: db}
}

// EventQuery defines filters for querying events.
type EventQuery struct {
	Type       *models.EventType  // Filter by event type
	EntityType *models.EntityType // Filter by entity type
	EntityID   *string            // Filter by entity ID
	Since      *time.Time         // Events at or after this time (inclusive)
	Until      *time.Time         // Events before this time (exclusive)
	Cursor     string             // Pagination cursor (event ID)
	Limit      int                // Max results to return
}

// EventPage represents a page of query results.
type EventPage struct {
	Events     []*models.Event
	NextCursor string
}

// Append adds a new event to the event log.
// Returns ErrInvalidEvent if required fields are missing.
func (r *EventRepository) Append(ctx context.Context, event *models.Event) error {
	if event.Type == "" || event.EntityType == "" || event.EntityID == "" {
		return ErrInvalidEvent
	}
	return r.Create(ctx, event)
}

// Create appends a new event to the event log.
func (r *EventRepository) Create(ctx context.Context, event *models.Event) error {
	return r.createWithExecutor(ctx, r.db, event)
}

// CreateWithTx appends a new event using an existing transaction.
func (r *EventRepository) CreateWithTx(ctx context.Context, tx *sql.Tx, event *models.Event) error {
	if tx == nil {
		return fmt.Errorf("transaction is required")
	}
	return r.createWithExecutor(ctx, tx, event)
}

func (r *EventRepository) createWithExecutor(ctx context.Context, execer eventExecer, event *models.Event) error {
	if event.Type == "" {
		return fmt.Errorf("event type is required")
	}
	if event.EntityType == "" {
		return fmt.Errorf("event entity type is required")
	}
	if event.EntityID == "" {
		return fmt.Errorf("event entity id is required")
	}

	if event.ID == "" {
		event.ID = uuid.New().String()
	}
	if event.Timestamp.IsZero() {
		event.Timestamp = time.Now().UTC()
	} else {
		event.Timestamp = event.Timestamp.UTC()
	}

	var payloadJSON *string
	if len(event.Payload) > 0 {
		s := string(event.Payload)
		payloadJSON = &s
	}

	var metadataJSON *string
	if event.Metadata != nil {
		data, err := json.Marshal(event.Metadata)
		if err != nil {
			return fmt.Errorf("failed to marshal metadata: %w", err)
		}
		s := string(data)
		metadataJSON = &s
	}

	_, err := execer.ExecContext(ctx, `
		INSERT INTO events (
			id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
		) VALUES (?, ?, ?, ?, ?, ?, ?)
	`,
		event.ID,
		event.Timestamp.Format(time.RFC3339),
		string(event.Type),
		string(event.EntityType),
		event.EntityID,
		payloadJSON,
		metadataJSON,
	)
	if err != nil {
		return fmt.Errorf("failed to insert event: %w", err)
	}

	return nil
}

// Get retrieves an event by ID.
func (r *EventRepository) Get(ctx context.Context, id string) (*models.Event, error) {
	row := r.db.QueryRowContext(ctx, `
		SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
		FROM events WHERE id = ?
	`, id)

	return r.scanEvent(row)
}

// Query retrieves events matching the given filters with cursor-based pagination.
func (r *EventRepository) Query(ctx context.Context, q EventQuery) (*EventPage, error) {
	limit := q.Limit
	if limit <= 0 {
		limit = 100
	}

	// Build query dynamically
	query := `SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json FROM events WHERE 1=1`
	args := []any{}

	if q.Type != nil {
		query += ` AND type = ?`
		args = append(args, string(*q.Type))
	}
	if q.EntityType != nil {
		query += ` AND entity_type = ?`
		args = append(args, string(*q.EntityType))
	}
	if q.EntityID != nil {
		query += ` AND entity_id = ?`
		args = append(args, *q.EntityID)
	}
	if q.Since != nil {
		query += ` AND timestamp >= ?`
		args = append(args, q.Since.UTC().Format(time.RFC3339))
	}
	if q.Until != nil {
		query += ` AND timestamp < ?`
		args = append(args, q.Until.UTC().Format(time.RFC3339))
	}
	if q.Cursor != "" {
		// Cursor is the last event ID; fetch events with timestamp >= cursor's timestamp
		// but exclude events with same timestamp and id <= cursor
		query += ` AND (timestamp, id) > (SELECT timestamp, id FROM events WHERE id = ?)`
		args = append(args, q.Cursor)
	}

	query += ` ORDER BY timestamp, id LIMIT ?`
	args = append(args, limit+1) // Fetch one extra to determine if there's a next page

	rows, err := r.db.QueryContext(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("failed to query events: %w", err)
	}
	defer rows.Close()

	var events []*models.Event
	for rows.Next() {
		event, err := r.scanEventFromRows(rows)
		if err != nil {
			return nil, err
		}
		events = append(events, event)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating events: %w", err)
	}

	page := &EventPage{}
	if len(events) > limit {
		// There's a next page
		page.Events = events[:limit]
		page.NextCursor = events[limit-1].ID
	} else {
		page.Events = events
	}

	return page, nil
}

// ListByEntity retrieves events for an entity, ordered by timestamp.
func (r *EventRepository) ListByEntity(ctx context.Context, entityType models.EntityType, entityID string, limit int) ([]*models.Event, error) {
	if limit <= 0 {
		limit = 100
	}

	rows, err := r.db.QueryContext(ctx, `
		SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
		FROM events
		WHERE entity_type = ? AND entity_id = ?
		ORDER BY timestamp
		LIMIT ?
	`, string(entityType), entityID, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query events: %w", err)
	}
	defer rows.Close()

	var events []*models.Event
	for rows.Next() {
		event, err := r.scanEventFromRows(rows)
		if err != nil {
			return nil, err
		}
		events = append(events, event)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating events: %w", err)
	}

	return events, nil
}

func (r *EventRepository) scanEvent(row *sql.Row) (*models.Event, error) {
	var event models.Event
	var timestamp, eventType, entityType string
	var payloadJSON sql.NullString
	var metadataJSON sql.NullString

	err := row.Scan(
		&event.ID,
		&timestamp,
		&eventType,
		&entityType,
		&event.EntityID,
		&payloadJSON,
		&metadataJSON,
	)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrEventNotFound
		}
		return nil, fmt.Errorf("failed to scan event: %w", err)
	}

	event.Type = models.EventType(eventType)
	event.EntityType = models.EntityType(entityType)

	if t, err := time.Parse(time.RFC3339, timestamp); err == nil {
		event.Timestamp = t
	}

	if payloadJSON.Valid {
		event.Payload = json.RawMessage(payloadJSON.String)
	}
	if metadataJSON.Valid {
		if err := json.Unmarshal([]byte(metadataJSON.String), &event.Metadata); err != nil {
			r.db.logger.Warn().Err(err).Str("event_id", event.ID).Msg("failed to parse event metadata")
		}
	}

	return &event, nil
}

func (r *EventRepository) scanEventFromRows(rows *sql.Rows) (*models.Event, error) {
	var event models.Event
	var timestamp, eventType, entityType string
	var payloadJSON sql.NullString
	var metadataJSON sql.NullString

	if err := rows.Scan(
		&event.ID,
		&timestamp,
		&eventType,
		&entityType,
		&event.EntityID,
		&payloadJSON,
		&metadataJSON,
	); err != nil {
		return nil, fmt.Errorf("failed to scan event: %w", err)
	}

	event.Type = models.EventType(eventType)
	event.EntityType = models.EntityType(entityType)

	if t, err := time.Parse(time.RFC3339, timestamp); err == nil {
		event.Timestamp = t
	}

	if payloadJSON.Valid {
		event.Payload = json.RawMessage(payloadJSON.String)
	}
	if metadataJSON.Valid {
		if err := json.Unmarshal([]byte(metadataJSON.String), &event.Metadata); err != nil {
			r.db.logger.Warn().Err(err).Str("event_id", event.ID).Msg("failed to parse event metadata")
		}
	}

	return &event, nil
}

// Count returns the total number of events.
func (r *EventRepository) Count(ctx context.Context) (int64, error) {
	var count int64
	err := r.db.QueryRowContext(ctx, `SELECT COUNT(*) FROM events`).Scan(&count)
	if err != nil {
		return 0, fmt.Errorf("failed to count events: %w", err)
	}
	return count, nil
}

// OldestTimestamp returns the timestamp of the oldest event.
func (r *EventRepository) OldestTimestamp(ctx context.Context) (*time.Time, error) {
	var timestamp sql.NullString
	err := r.db.QueryRowContext(ctx, `SELECT MIN(timestamp) FROM events`).Scan(&timestamp)
	if err != nil {
		return nil, fmt.Errorf("failed to get oldest timestamp: %w", err)
	}
	if !timestamp.Valid {
		return nil, nil // No events
	}
	t, err := time.Parse(time.RFC3339, timestamp.String)
	if err != nil {
		return nil, fmt.Errorf("failed to parse oldest timestamp: %w", err)
	}
	return &t, nil
}

// DeleteOlderThan deletes events older than the given timestamp.
// Returns the number of events deleted.
func (r *EventRepository) DeleteOlderThan(ctx context.Context, before time.Time, limit int) (int64, error) {
	if limit <= 0 {
		limit = 1000
	}

	result, err := r.db.ExecContext(ctx, `
		DELETE FROM events WHERE id IN (
			SELECT id FROM events WHERE timestamp < ? ORDER BY timestamp LIMIT ?
		)
	`, before.UTC().Format(time.RFC3339), limit)
	if err != nil {
		return 0, fmt.Errorf("failed to delete old events: %w", err)
	}

	count, err := result.RowsAffected()
	if err != nil {
		return 0, fmt.Errorf("failed to get deleted count: %w", err)
	}
	return count, nil
}

// DeleteExcess deletes the oldest events beyond a maximum count.
// Returns the number of events deleted.
func (r *EventRepository) DeleteExcess(ctx context.Context, maxCount int, limit int) (int64, error) {
	if maxCount <= 0 {
		return 0, nil
	}
	if limit <= 0 {
		limit = 1000
	}

	// Count current events
	total, err := r.Count(ctx)
	if err != nil {
		return 0, err
	}

	excess := total - int64(maxCount)
	if excess <= 0 {
		return 0, nil
	}

	// Limit to batch size
	deleteCount := excess
	if deleteCount > int64(limit) {
		deleteCount = int64(limit)
	}

	result, err := r.db.ExecContext(ctx, `
		DELETE FROM events WHERE id IN (
			SELECT id FROM events ORDER BY timestamp LIMIT ?
		)
	`, deleteCount)
	if err != nil {
		return 0, fmt.Errorf("failed to delete excess events: %w", err)
	}

	count, err := result.RowsAffected()
	if err != nil {
		return 0, fmt.Errorf("failed to get deleted count: %w", err)
	}
	return count, nil
}

// ListOlderThan retrieves events older than the given timestamp, ordered by timestamp.
// Used for archiving before deletion.
func (r *EventRepository) ListOlderThan(ctx context.Context, before time.Time, limit int) ([]*models.Event, error) {
	if limit <= 0 {
		limit = 1000
	}

	rows, err := r.db.QueryContext(ctx, `
		SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
		FROM events
		WHERE timestamp < ?
		ORDER BY timestamp
		LIMIT ?
	`, before.UTC().Format(time.RFC3339), limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query old events: %w", err)
	}
	defer rows.Close()

	var events []*models.Event
	for rows.Next() {
		event, err := r.scanEventFromRows(rows)
		if err != nil {
			return nil, err
		}
		events = append(events, event)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating old events: %w", err)
	}

	return events, nil
}

// ListOldest retrieves the oldest events up to the given limit.
// Used for archiving excess events before deletion.
func (r *EventRepository) ListOldest(ctx context.Context, limit int) ([]*models.Event, error) {
	if limit <= 0 {
		limit = 1000
	}

	rows, err := r.db.QueryContext(ctx, `
		SELECT id, timestamp, type, entity_type, entity_id, payload_json, metadata_json
		FROM events
		ORDER BY timestamp
		LIMIT ?
	`, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query oldest events: %w", err)
	}
	defer rows.Close()

	var events []*models.Event
	for rows.Next() {
		event, err := r.scanEventFromRows(rows)
		if err != nil {
			return nil, err
		}
		events = append(events, event)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating oldest events: %w", err)
	}

	return events, nil
}

// DeleteByIDs deletes events by their IDs.
// Returns the number of events deleted.
func (r *EventRepository) DeleteByIDs(ctx context.Context, ids []string) (int64, error) {
	if len(ids) == 0 {
		return 0, nil
	}

	// Build placeholder list
	placeholders := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		placeholders[i] = "?"
		args[i] = id
	}

	query := fmt.Sprintf(`DELETE FROM events WHERE id IN (%s)`, strings.Join(placeholders, ","))
	result, err := r.db.ExecContext(ctx, query, args...)
	if err != nil {
		return 0, fmt.Errorf("failed to delete events by ids: %w", err)
	}

	count, err := result.RowsAffected()
	if err != nil {
		return 0, fmt.Errorf("failed to get deleted count: %w", err)
	}
	return count, nil
}
